### DISCUSSION TITLE
What is the deno equivalent to "npm link" ?

### DISCUSSION THREAD

**Discussion by justinmchase** [denoland/deno#11425] [Q&A]
Suppose I have two repos, one depends on the other `A -> B`. I'm working on a feature in `B` and testing it with code in `A` before publishing.

What is the best way to create a local only link from `A` to the local folder containing the code of `B`? Ideally `A` would still have a dependency on `https://deno.land/x/B/mod.ts` in the deps.ts file but not necessarily.

With npm you would cd into B and run `npm link` and that essentially registers that module globally and then cd into A and run `npm link B` and it would use symlinks to do some `node_modules` trickery.

Other projects like [lerna](https://www.npmjs.com/package/lerna) allow you to have multiple packages in a single repo and then you can configure relationships between them and it automatically links them all together with a single command.

So I'm just wondering what the equivalent in Deno is? Whats the recommended way to temporarily link modules in local folders together?

I was thinking of using the [import map](https://deno.land/manual/linking_to_external_code/import_maps), but sort of in reverse? Though I'm not sure if it would work... but basically:

```ts
// link.json
{
  "imports": {
    // "fmt/": "https://deno.land/std@0.101.0/fmt/"
    "https://deno.land/example@0.0.1/": "../example/"
   }
}
```

Then when I run locally:
```sh
deno --import-map=link.json main.ts
```

I'm not sure if that would actually work though.

**Comment by bartlomieju** [comment]
> I was thinking of using the import map, but sort of in reverse? Though I'm not sure if it would work... but basically:

That's exactly the solution. Your example is correct

**Reply by justinmchase** [reply]
Ok, I thought of it while I was writing it up and decided to post anyway 😝 Hopefully it will help someone in the future.

**Reply by justinmchase** [reply]
I just searched for this again in google and it landed me on this post, which I completely forgot about writing. So that future someone turns out to be myself, lol.

**Reply by jollytoad** [reply]
I'm working on a Deno project that has several separate repos, and have two import maps (import_map.json & import_map_local.json) where the local one has relative paths to the other repos. I can then just switch the import map reference in the deno.json when i need to test against the local stuff. No need for npm link or mono-repo faff, Deno makes this such a joy to work with.

**Reply by justinmchase** [reply]
That is basically what I did too, added something like this in my deno.json:

```ts
{
  "tasks": {
    "run": "deno run main.ts",
    "run:link": "deno run --import-map=link.json main.ts"
  }
}
```

I only have one repo to link though so its pretty easy but in the case where you have multiple related repos, how do you manage linking one but not others?

**Reply by jollytoad** [reply]
I generally have all repos checked out at the main branch (not third-party repos), so they are usually equivalent to the releases, so isn't really an issue. Occasionally i'll just duplicate the main import map and hack individual entries if necessary, esp if I want to try out a third-party repo locally. Embrace variant import maps :D ... generally I find it a lot simpler doing this than the old link faff, I used to have so many problems with npm/yarn linking, was never sure if they were doing what I thought they should be doing.

**Reply by justinmchase** [reply]
> ...was never sure if they were doing what I thought they should be doing.

True story.
